### ISSUE TITLE
How to use AMS plug-ins to build modular synth in Synthpod?

### ISSUE THREAD

**Issue by unfa** [OpenMusicKontrollers/synthpod#46]
I'm trying to create the simplest modular synthesizer in Synthpod.

I'm having trouble getting a Gate from my MIDI input signal to trigger the AMS Envelope.

Is there some helper plug-in that does this?

Have anyone made working modular synths in Synthpod? I'm hyped for that, as being able to do modular synthesis in a Plug-in inside Ardour (or other DAW) is a dream come true...

**Comment by ventosus** [comment]
Ah, synthpod cannot automate CV inputs/outputs directly, yet.

but ams-lv2 has a plugin called control2cv, you can use that until this feature is added.

**Comment by unfa** [comment]
I think I need a plugin that'll generate gate CV from MIDI note inputs, does control2cv do that?

**Comment by ventosus** [comment]
No, but control ports in synthpod can be automated with MIDI CC and OSC. There are plugins from x42 midi-filters that can convert MIDi Note to MIDI CC.

**Comment by ventosus** [comment]
So, this may work:

* Synthpod MIDI in
* http://gareus.org/oss/lv2/midifilter#notetocc
* http://github.com/blablack/ams-lv2/controltocv
* http://github.com/blablack/ams-lv2/env

**Comment by unfa** [comment]
Yes, I was able to manually trigger an AMS envelope by moving eh Gate knob in the Synthpod UI - but getting this from notes is a must.

I'll try the x42 plug-in - thanks!

**Comment by PrinzNika** [comment]
commenting to state my interest in the same topic.
I dont really understand how this works, but it seems like a rock solid application.

I am just wondering how to automate these parameters that you can click on when the node is selected. Or does selecting the parameters not do anything yet? I m wondering how to open an output for a given parameter of a node like when in ingen you connect a saw out of an lfo to the freq in if an oscillator. in that programm all the possible parameters are listed within the node. I am guessing that this is achieved with the patchmatrix like node that appearrs when you click two nodes together?

**Comment by ventosus** [comment]
> I dont really understand how this works, but it seems like a rock solid application.

It is not, of course, still in a developmental stage.

> I am just wondering how to automate these parameters that you can click on when the node is selected

Select the the parameter (right click), and an automation tab will pop up. There you can create input/output automations for controls only.

Synthpod can route 3 types of signals:
* audio
* CV
* (atom)events (MIDI, OSC, ...)

Each plugin in synthpod gets an extra pair of (atom)event ports (automation_in, automation_out). When you automate a control, this will generate events (MIDI, OSC) on those two (atom)event ports. They can be routed like any other (atom)event port.

As stated in this issue, CV ports cannot be automated directly, yet. You'll need to create the CV signal with some other plugin (look above for a possible approach).

**Comment by PrinzNika** [comment]
Ok I had some time to quickly give it a look and from that quick look, 3 questions emerged.
In the VM plugin bundle, I dont quite understand the order of operation. I can successfully input signals and with the default config the same signals get routed to the output. But I dont understand what operation goes to which channel in which order.
So if you could help me out with the following scenario I think I could figure out the rest.
Lets take to inputs. Both sine waves with different frequencies. on output 1 should be the first input, on output 2 the second. output 3 lets say we add both waves together. output 4 divides the signal on output 3 with a cos function. With this example I think I could understand the general flow of operations.

In Synthpod, when I load a plugin, I dont reeally know how to automate a parameter. If I load Synthv1 in Ingen, I get a Node with all the control inputs.
When I load te same plugin in Synthpod, there is no input in the cv tab, even with the automation option enabled in the top bar. If I right click the plugin, there is a list of parameters, when I click on automate I can select a midi or osc channel. However I cant figure out how to use a signal, generated by another node on the canvas. How would I do that? I realized that these control inputs in Ingen arent cv signal, but control. But still I cant find a way to automate the parameters shown in the controls tap with somthing sent out by another node.
I mean yes I got to a point where I found some automation In/ Automation out connections, but the again I dont know how to assign the automation to a specific control
Basically I didnt understand the basic signal routing workfolw just yet.
And the last thing is, that my plugins are missing the information about which bundle they are in. My bundle tab is empty in the window on the bottom left, when i activate the settings tab.

With all my questions you always have to consider, that I am still a total noob so please forgive me when I ask something obvious, that requires only the basic knowledge.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by ventosus** [comment]
>In Synthpod, when I load a plugin, I dont reeally know how to automate a parameter. If I load Synthv1 in Ingen, I get a Node with all the control inputs.

synthpod only routes sample-based plugin ports (AUDIO, CV) and event-based plugin ports (MIDI, OSC, TIME, ATOMs).

There is a difference between a CONTROL and a CV port. A CONTROL changes only once per period (e.g. only every 128 samples), a CV port - just like AUDIO - changes every sample.

> When I load te same plugin in Synthpod, there is no input in the cv tab, even with the automation option enabled in the top bar. If I right click the plugin, there is a list of parameters, 

synthpod does expose CV ports on the patch canvas. CONTROL's are not exposed on the patch canvas, but on the properties tab, because it does not make any sense to route those.

> I mean yes I got to a point where I found some automation In/ Automation out connections, but the again I dont know how to assign the automation to a specific control
Basically I didnt understand the basic signal routing workfolw just yet.

Automation in synthpod is done solely by event messaging.

Every plugin in synthpod gets two additional event ports (automation in/out). If you create an output automation for CONTROL X, whenever that CONTROL changes, an event (MIDI, OSC) will be sent out of the automation_out port. Just route the automation_out port to an automation_in port of another plugin. There you create an input automation with the same MIDI/OSC message, and done you are.

I hope this make more sense now.

> And the last thing is, that my plugins are missing the information about which bundle they are in. My bundle tab is empty in the window on the bottom left, when i activate the settings tab.

Those are synthpod bundles, not plugin bundles, if you have not saved any synthpod projects ini ~/.lv2, this list will be empty, indeed.

**Comment by PrinzNika** [comment]
Aha ok I didnt know before that these control changes were native to some platform. Thanks for your explanation it makes more sense now.

Can you link me to documentation of your plugin bundles? Or is it somewhat essential for the use of them to know a scripting lanuage like lua?
‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by PrinzNika** [comment]
Awesome thanks for the elaborations!
About your plug-in bundles. I'm trying to figure the out as well, because I'm thinking that they are somewhat essential to the type of workflow suggested in synthpod.
Could you elaborate on VM? I sense that it has the potential to become my main lfo generating unit (even though I couldn't quite understand it's usuage yet).
What is the order of operations?
For me it is applied from bottom to to. So the lowest operation entry actually defines the the first output.
So far I realized that when I insert for example the '+' - (add) operation before (ie below) something, it adds the next 2(above) together.
It feels like inserting this '+'operation would invoke an add function that has two variables, which are automatically filled with the preceding 2 statements (the ones that are above the '+')
Can you elaborate on how I could
°output an lfo synced to a clock
°perform mathematical operations on an lfo
°understand the functionality behind it so I can at one point code automation curves in moony

Thanks for your help
-------- Original-Nachricht --------
An 18. Okt. 2019, 22:18, Hanspeter Portner schrieb:

**Comment by ventosus** [comment]
> So far I realized that when I insert for example the '+' - (add) operation before (ie below) something, it adds the next 2(above) together.

That's pretty much it, yes, it is called reverse polish notation [1].

[1] https://en.wikipedia.org/wiki/Reverse_Polish_notation

**Comment by PrinzNika** [comment]
Ahja nice now I'm able to use it!
Maybe consider mentioning it in the description on the website (unless this is a well known standart in programming that I just don't know of since I'm still a noob)
-------- Original-Nachricht --------
An 20. Okt. 2019, 14:32, Hanspeter Portner schrieb:

**Comment by PrinzNika** [comment]
Can I add a feature request?

It's be lovely to make notes on the canvas

-------- Original-Nachricht --------
An 20. Okt. 2019, 14:32, Hanspeter Portner schrieb:

**Comment by ventosus** [comment]
> Can I add a feature request?

Sure, but:

So, please open a new issue for everything that does not strictly belong into this issue here (e.g. also the questions about vm.lv2 do not belong here) !

This has the advantage that issues will not be forgotten and other people looking for the same information will easily find it, if placed in a separate issue ticket int the right project.

**Comment by PrinzNika** [comment]
Oh sorry, I had actually answering via email and didn't realize that it would get posted into the GitHub thread

-------- Original-Nachricht --------
An 23. Okt. 2019, 12:37, Hanspeter Portner schrieb:

**Comment by falkTX** [comment]
Adding to the discussion here, mod-cv-plugins were released recently, they provide a monophonic and polyphonic "MIDI 2 CV" plugin.
See https://github.com/moddevices/mod-cv-plugins

**Comment by grammoboy2** [comment]
FYI I got Non-Timeline's CV automation working to apply automation to standalone Zynaddsubfx using synthpod and ams-lv2: https://github.com/grammoboy2/non-zyn-CV

**Comment by PrinzNika** [comment]
would anyone here mind sharing some patches that further illustrate the use and workflow practices of synthpod? Im still struggeling to use it to its potential

**Comment by ventosus** [comment]
closing this, as it is not really an issue in the first place.
